### ISSUE TITLE
onSetActive, onSetInactive callbacks firing incorrectly or not at all

### ISSUE THREAD

**Issue by nemanjam** [fisshy/react-scroll#450]
If I omit `containerId` prop on `Link` component and let default `document` scroll container I get incorrectly highlighted navbar item (previous item is still highlighted).

If I pass `id` of the div that has scroll to `Link` then `onSetActive`, `onSetInactive` callbacks aren't firing at all.

Also if I specify `offset` prop, scrolling position is correct but highlighted navitem is incorrect (`onSetActive`, `onSetInactive` callbacks).

I tried setting a breakpoint at i.e.

```
webpack:///node_modules/react-scroll/modules/mixins/scroll-link.js

line 264 _this2.props.onSetInactive && _this2.props.onSetInactive(to, element);
```

and it's not pausing at all. I tried investigating stack trace when `container` is document and it does pause in debugger but I can just see handlers array gets invoked and before that mount is called, it's not much informative, I expected some conditions and then to see which one from them isn't fulfilled.

To summarize: 

`onSetActive`, `onSetInactive` callbacks aren't firing at all when `containerId` is passed, or firing incorrectly when `containerId` is omitted or `offset` prop is being used.

Is this a bug, and is there a workaround for this? Can you give me advice what's the best way to debug this?

Also I noticed `container: PropTypes.object,` isn't supported in types in `"@types/react-scroll": "^1.8.2",`:

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react-scroll/modules/components/Link.d.ts

**Comment by fisshy** [comment]
Do you have a plunkr/codepen demonstrating you problem? 

Did you also set the property spy={true}?

**Comment by fisshy** [comment]
If you look at this codesandbox and check the console.log output, you can see that the onSetActive/onSetInactive is handeled correctly.

https://codesandbox.io/s/basic-forked-20jhg?file=/src/index.js

**Comment by nemanjam** [comment]
Here it is, I set `containerId="html-tag-id"` on first 4 `Link` and they stopped logging. I also set `<html id="html-tag-id" lang="en">` in `public/index.html` because `html` tag has scroll.

https://codesandbox.io/s/basic-forked-nggx8?file=/src/index.js

**Comment by therealdrtroll** [comment]
@fisshy I am facing the same issue but only for larger screen sizes(24 inch monitor) . Works well for smaller screen sizes (macbook 13 inch).

**Comment by fisshy** [comment]
@therealdrtroll Could you provide a plunkr? I believe your "elements" are too small to trigger the spy when scrolling?

**Comment by therealdrtroll** [comment]
@fisshy Please check through this codesandbox link
link: https://codesandbox.io/s/react-scroll-active-issue-o3h5k?file=/src/App.js

* If you got to top(start position) and try to click a tab, it doesn't highlight correctly, But once you are on an element and click again it works fine. 
* Try going to top again and click on any tab it will be incorrect again.
* If we remove the animation (smooth = {false}) it works fine.

**Comment by fisshy** [comment]
@therealdrtroll Your problem is the

```
.active {
  color: white;
  background-color: red;
  /*padding: 6px 5px;*/
  border-radius: 8px;
}
```

First you click the link, then it calculate the position of the element, but once active class hits a Link, then the actual element gets pushed a few pixels below the actual scroll position.

Add the padding to your links from start or remove the padding from .active

**Comment by therealdrtroll** [comment]
@fisshy Got it. Thank you for your time.

**Comment by fisshy** [comment]
@therealdrtroll Glad I could help :)

**Comment by rgurgul** [comment]
<link offset={50}...  should help
